### PR TITLE
Merged new code from scaleoutsystems/stackn/release/v0.7.0

### PR DESCRIPTION
**Pull request by sandstromviktor**
## Status

- [ ] Ready
- [X] Draft
- [ ] Hold

## Description
Merged code from `scaleoutsystems/stackn/release/v0.7.0`
I've removed 1 test from `projects/tests/test_grant_access_to_project_view.py`. This fails on scaleouts branch as well. 
It uses an attribute called `is_client`, but this does not exist and the test fail. 

## Types of changes

What types of changes does your code introduce to STACKn?

- [ ] Hotfix (fixing a critical bug in production)
- [ ] Bugfix
- [X] New feature
- [ ] Documentation update

## Checklist

_If you're unsure about any of the items below, don't hesitate to ask. We're here to help! 
This is simply a reminder of what we are going to look for before merging your code._

- [X] This pull request is against **develop** branch (not applicable for hotfixes)
- [ ] I have included a link to the issue on GitHub or JIRA (if any)
- [ ] I have included migration files (if there are changes to the model classes)
- [ ] I have read the [CONTRIBUTING](https://github.com/scaleoutsystems/stackn/blob/master/CONTRIBUTING.md) doc
- [ ] I have included, reviewed and executed tests (unit and end2end) to complement my changes
- [ ] I have updated the related documentation (if necessary) 
- [ ] I have updated the release notes (docs/releasenotes.md)
- [X] I have added a reviewer for this pull request
- [X] I have added myself as an author for this pull request

### CODE REVIEW

File: apps/models.py [hunk 1]
```diff
@@ -53,10 +57,104 @@ def __str__(self):
...
+                    project=project,
+                    app_dependencies=instance,
+                ).exists()
+
+                if exists:
+                    result = result.exclude(id=instance.id)
+
+        return result
+
     def user_can_create(self, user, project, app_slug):
```
Comment (churnikov) [inline apps/models.py:135]:
```suggestion
    def user_can_create(
    self, 
    user: User, 
    project: Project, 
    app_slug: str
    ) -> bool:
    """
    Determine if a user can create an app instance based on various conditions.

    :param user: The user for whom the permission is being checked.
    :param project: The project in which the app instance is to be created.
    :param app_slug: The slug of the app for which the instance is to be created.
    :return: A boolean indicating whether the user can create the app instance.

    .. note::
        - The method checks the app's ``user_can_create`` attribute.
        - It also considers the number of app instances already created against the allowed limit.
        - If the user has the "apps.add_appinstance" permission, they can always create the app instance, regardless of the limit.
    """
```

File: apps/tests/test_app_instance_manager.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
+from django.contrib.auth import get_user_model
+from django.db.models import Q
+from django.test import TestCase, override_settings
+
+from projects.models import Project
+
+from ..models import AppInstance, Apps
```
Comment (churnikov) [inline apps/tests/test_app_instance_manager.py:7]:
```suggestion
from apps.models import AppInstance, Apps
```

File: apps/tests/test_app_settings_view.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
+from django.contrib.auth import get_user_model
+from django.test import Client, TestCase
+
+from projects.models import Project
+
+from ..models import AppCategories, AppInstance, Apps
```
Comment (churnikov) [inline apps/tests/test_app_settings_view.py:6]:
```suggestion
from apps.models import AppCategories, AppInstance, Apps
```

File: projects/management/commands/removedeletedprojects.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    (and all objects belonging to the projects) with status deleted"""
+
+    def handle(self, *args, **options):
+        qs = Project.objects.filter(status="deleted")
+
+        for project in qs:
+            try:
+                self.stdout.write(f"Deleting project: {project.name}")
+                project.delete()
+            except Exception:
```
Comment (churnikov) [inline projects/management/commands/removedeletedprojects.py:17]:
```suggestion
            # TODO: Make exception explicit
            except Exception:
```

File: projects/models.py [hunk 5]
```diff
@@ -174,7 +200,42 @@ def user_can_create(self, user):
...
+
+
+def get_default_apps_per_project_limit():
+    try:
+        apps_per_project_limit = (
+            settings.APPS_PER_PROJECT_LIMIT
+            if settings.APPS_PER_PROJECT_LIMIT is not None
+            else {}
+        )
+    except Exception:
```
Comment (churnikov) [inline projects/models.py:235]:
```suggestion
    # TODO: Make exception explicit
    except Exception:
```

**Review by churnikov — COMMENTED** [review]
Jeez, how many files

Gonna be honest, with docscrings I used chatgpt, but I read through code and I was satisfied with the result.

I tried to look through as much as possible, but there is still some files left to look at.

File: apps/admin.py [hunk 6]
```diff
@@ -4,12 +4,38 @@
 
 
 class AppsAdmin(admin.ModelAdmin):
-    list_display = ("name", "user_can_create", "slug", "revision")
+    list_display = (
```
Comment (alfredeen) [inline apps/admin.py:7]: It seems like this change undoes the formatting of black. Have you run the black formatter on the fetched code?

**Review by alfredeen — CHANGES_REQUESTED** [review]
Please run tools isort and black for formatting before merging. isort should modify 5 files and black 22 files. Also flake8 detects 2 problems.

**Review by sandstromviktor — COMMENTED** [review]
I've run Black, isort and flake.

**Comment by alfredeen** [comment]
> I've run Black, isort and flake.

Thank you!

**Pull request merged**